### ISSUE TITLE
Coveralls fails to upload with HTTP 422 "Unprocessable Entity" on retried builds

### ISSUE THREAD

**Issue by nolar** [lemurheavy/coveralls-public#1435]
Hello.  First of all, thank you for your service free of charge! 

Now, to the problem:

Coveralls fails on sending the data with HTTP 422 Unprocessable Entity:

* https://travis-ci.org/github/zalando-incubator/kopf/jobs/680475472

```
$ coveralls
Submitting coverage to coveralls.io...
Could not submit coverage: 422 Client Error: Unprocessable Entity for url: https://coveralls.io/api/v1/jobs
Traceback (most recent call last):
  File "/home/travis/virtualenv/python3.7.1/lib/python3.7/site-packages/coveralls/api.py", line 194, in wear
    response.raise_for_status()
  File "/home/travis/virtualenv/python3.7.1/lib/python3.7/site-packages/requests/models.py", line 941, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 422 Client Error: Unprocessable Entity for url: https://coveralls.io/api/v1/jobs
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/home/travis/virtualenv/python3.7.1/lib/python3.7/site-packages/coveralls/cli.py", line 78, in main
    result = coverallz.wear()
  File "/home/travis/virtualenv/python3.7.1/lib/python3.7/site-packages/coveralls/api.py", line 197, in wear
    raise CoverallsException('Could not submit coverage: {}'.format(e))
coveralls.exception.CoverallsException: Could not submit coverage: 422 Client Error: Unprocessable Entity for url: https://coveralls.io/api/v1/jobs
The command "coveralls" exited with 1.
```

This happens always with one specific scenario:

* Run a build normally, allow the tests to fail, allow the coverage to upload whatever it has collected.
* Retrigger the build.
* The tests do pass now (it was a temporary timing error).
* The coverage upload now does not work with HTTP 422.

First of all, some better error message or status code would be nice. HTTP 409 "Conflict" is usually returned when trying to create something that already exists (if that is actually the cause).

And second: how to fix this issue? Previously, it worked even when the tests were failing and then retried.

Thanks.

**Comment by Ark-kun** [comment]
We have the same issue in many of our PRs. We use coveralls in Travis. Couple of days ago it stopped working in some cases. And the retries do not help.
I'll be removing coveralls for now to unblock the PRs.

**Comment by piyush0** [comment]
I had the same issue, make sure that travis has `COVERALLS_REPO_TOKEN` this environment variable correctly set.

**Comment by hannes-ucsc** [comment]
Yes, status code should be 409.

**Comment by duttonw** [comment]
un-related but still helpful for me.

If you fork a repo that has circleci and it uses coveralls, you need to setup coveralls yourself and add to the repo project environment settings the COVERALLS_REPO_TOKEN  and the token for your forked repo.

I think this is the case if "Pass secrets to builds from forked pull requests" is set to Disabled.

**Comment by solancer** [comment]
I was facing the same issue on Github actions, adding `--service=gihtub` in front of the coveralls command fixed it for me

- name: Coveralls
      env:
        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
      run: coveralls --service=github

**Comment by jugmac00** [comment]
> I was facing the same issue on Github actions, adding `--service=gihtub` in front of the coveralls command fixed it for me
> 
> * name: Coveralls
>   env:
>   GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
>   run: coveralls --service=github

Same issue - and this solution is not helping.

I mirgrated 20+ repositories from Travis CI to GitHub Actions, and of those ca. 17 just worked, and 3 do not work at all and error with above (422) error message, although all repositories have a uniform layout.

gh actions with errors
https://github.com/morepath/morepath_rest_dump_load/runs/1754287368?check_suite_focus=true
https://github.com/morepath/morepath_batching/actions/runs/497059413
https://github.com/morepath/more.mako/runs/1730894898?check_suite_focus=true

**Comment by gdubicki** [comment]
>       run: coveralls --service=github

Thanks @solancer , this helped.

But I was really close to not even trying this. Because a moment before Coveralls outputted this in a previously failed run:
```
coveralls.exception.CoverallsException: Running on Github Actions but GITHUB_TOKEN is not set. Add "env: GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}" to your step config.
```
...so I was sure that it already *knows* that it is running on GitHub (Actions).

**Comment by solancer** [comment]
> > I was facing the same issue on Github actions, adding `--service=gihtub` in front of the coveralls command fixed it for me
> > 
> > * name: Coveralls
> >   env:
> >   GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
> >   run: coveralls --service=github
> 
> Same issue - and this solution is not helping.
> 
> I mirgrated 20+ repositories from Travis CI to GitHub Actions, and of those ca. 17 just worked, and 3 do not work at all and error with above (422) error message, although all repositories have a uniform layout.
> 
> gh actions with errors
> https://github.com/morepath/morepath_rest_dump_load/runs/1754287368?check_suite_focus=true
> https://github.com/morepath/morepath_batching/actions/runs/497059413
> https://github.com/morepath/more.mako/runs/1730894898?check_suite_focus=true

Try setting the env before the run

**Comment by jugmac00** [comment]
@solancer I am not sure what you mean with `before the run`.

Could you create a Pull Request to any of the three mentioned repositories?

That would be super awesome!

**Comment by afinetooth** [comment]
@jugmac00 can you please try running some builds in debug mode for those failing repos, per [instructions here](https://coveralls-python.readthedocs.io/en/latest/troubleshooting.html).

Please share the new build logs and we'll see if we can gain some more insight as to what is failing and why.

**Comment by solancer** [comment]
@jugmac00 I've raised a PR with a fix, seems to work fine now

https://github.com/morepath/morepath_rest_dump_load/pull/12

**Comment by jugmac00** [comment]
@afinetooth As requested, I ran coveralls with the debug command, see https://github.com/morepath/morepath_rest_dump_load/pull/14/checks?check_run_id=1767517263

The relevant output is:
```
Run pip install coveralls
Collecting coveralls
  Downloading coveralls-3.0.0-py2.py3-none-any.whl (13 kB)
Collecting requests>=1.0.0
  Downloading requests-2.25.1-py2.py3-none-any.whl (61 kB)
Collecting docopt>=0.6.1
  Downloading docopt-0.6.2.tar.gz (25 kB)
Collecting coverage<6.0,>=4.1
  Downloading coverage-5.3.1-cp39-cp39-manylinux2010_x86_64.whl (244 kB)
Collecting certifi>=2017.4.17
  Downloading certifi-2020.12.5-py2.py3-none-any.whl (147 kB)
Collecting chardet<5,>=3.0.2
  Downloading chardet-4.0.0-py2.py3-none-any.whl (178 kB)
Collecting idna<3,>=2.5
  Downloading idna-2.10-py2.py3-none-any.whl (58 kB)
Collecting urllib3<1.27,>=1.21.1
  Downloading urllib3-1.26.2-py2.py3-none-any.whl (136 kB)
Building wheels for collected packages: docopt
  Building wheel for docopt (setup.py): started
  Building wheel for docopt (setup.py): finished with status 'done'
  Created wheel for docopt: filename=docopt-0.6.2-py2.py3-none-any.whl size=13705 sha256=a49c0db692e5e04e9eb0f135e93357271954c5786cb44a1a89f3d8069c65a8e4
  Stored in directory: /home/runner/.cache/pip/wheels/70/4a/46/1309fc853b8d395e60bafaf1b6df7845bdd82c95fd59dd8d2b
Successfully built docopt
Installing collected packages: urllib3, idna, chardet, certifi, requests, docopt, coverage, coveralls
Successfully installed certifi-2020.12.5 chardet-4.0.0 coverage-5.3.1 coveralls-3.0.0 docopt-0.6.2 idna-2.10 requests-2.25.1 urllib3-1.26.2
Missing .coveralls.yml file. Using only env variables.
Testing coveralls-python...
{"source_files": [], "git": {"branch": "run-coveralls-in-debug-mode", "head": {"id": "746233dd8b1ad0854d83f88cff871f34ffd1f4ed", "author_name": "J\u00fcrgen Gmach", "author_email": "juergen.gmach@googlemail.com", "committer_name": "J\u00fcrgen Gmach", "committer_email": "juergen.gmach@googlemail.com", "message": "Run coveralls integration in debug mode"}, "remotes": [{"name": "origin", "url": "https://github.com/morepath/morepath_rest_dump_load"}]}, "service_job_id": null, "repo_token": "[secure]", "service_name": "github", "service_number": "511454775", "config_file": ".coveragerc"}
==
Reporting 0 files
==
```

The corresponding `coveralls.io` page is https://coveralls.io/github/morepath/morepath_rest_dump_load which still shows "Here are some quick instructions for this Python project..." and not showing any successful report.

@solancer Thanks so much for your effort. Unfortunately, you also seem to have activated the debug mode, which means, it is not tried to send data to coveralls, and thus the build succeeds, but without reporting :-/

**Comment by afinetooth** [comment]
@jugmac00 yes, the issue is that no source files are being sent. 

Per the JSON in the debug output, the `source_files` hash is `empty` (`[]`):

```
{"source_files": [] [...] }
```

This means you'll need to configure your project to pass your coverage report to your python-based coveralls integration, so it can convert the report into JSON and POST it to the coveralls API.

As an example, here are instructions for [coveralls-python](https://github.com/TheKevJames/coveralls-python):
https://coveralls-python.readthedocs.io/en/latest/usage/configuration.html#configuration

**Comment by jugmac00** [comment]
@afinetooth Thanks for you suggestion.

I have read the documentation but found no clue what I should change.

See, I have 20 repositories with the almost exact same configuration, 17 work, 3 fail.

I just copied the 100% exact same configuration (both gh action and tox config) from https://github.com/morepath/morepath_sqlalchemy and created a PR for https://github.com/morepath/morepath_rest_dump_load/pull/15 - the first one works, the latter one fails.

I cannot imagine this is a configuration issue.

**Comment by afinetooth** [comment]
Hi @jugmac00 I still think the underlying issue here is that the project is not sending any source files to the coveralls API (according to that debug output). 

```
{"source_files": [], "git": {"branch": "run-coveralls-in-debug-mode", [...]
```

"Configuration" was probably not the clearest description of what needs to be done. What the empty `source_files` array means, most likely, is that for some reason the coveralls-python integration is not finding the coverage report your code is generating. (Thus, not populating `source_files` in the JSON.)

coveralls-python uses a default location where it looks for your coverage report. If it's not finding it, it means your project might be failing to generate one, or that it's saving it someplace besides the standard location.

[This documentation](https://coveralls-python.readthedocs.io/en/latest/usage/configuration.html#configuration) gives the config flag to tell coveralls-python where your coverage report is:

```
coveralls --rcfile=/path/to/coveragerc
```

You'll want to find out:

- If your project is indeed generating a coverage report
- Where it's saving that report
- If it's in a format that coveralls-python can handle

Once you do, you can (hopefully) use the flag above to pass the right location.

Here are two things to try in the meantime to help get you there:

1. Run one of your successful projects in debug mode. This will validate we have the right issue by showing a full `source_files` array in the debug output.
2. If that works, find out where the successful project is storing its coverage report; then see if you can determine any difference in the non-working project as to where it's storing its report.

**Comment by zgoda** [comment]
I'm getting this failure if action is triggered by github pull request. If i rerun action jobs for the very same PR manually - everything is fine. Actions triggered by push (eg after merging the before mentioned PR) are also doing fine.

**Comment by afinetooth** [comment]
@zgoda if your project is private and you're concerned about security, you can send details to support@coveralls.io and reference [this issue](https://github.com/lemurheavy/coveralls-public/issues/1435#issuecomment-820678511).

We would need to see your Github Actions workflow file, programming language and coveralls integrations you're using.

**Comment by zgoda** [comment]
@afinetooth eg this one https://github.com/zgoda/kristall/blob/master/.github/workflows/tests.yml, issue observed with this PR https://github.com/zgoda/kristall/pull/3. No trace in actions because I re-ran all action jobs manually and they succeeded. The sequence was like:

1. PR open by dependabot
2. job where coverage is collected fail on coveralls step because of missing `GITHUB_TOKEN`
3. update workflow in main branch including the above token, job w/ coverage succeeds
4. PR rebased
5. job fails on coveralls step because of 422 Unprocessable Entity
6. re-run manually all jobs, succeed
7. merge PR, all jobs succeed

Tried to reproduce with simple PR but now it works fine.

[Project on Coveralls](https://coveralls.io/github/zgoda/kristall)

**Comment by zgoda** [comment]
Okay, got a repro.

PR: https://github.com/zgoda/flask-rollup/pull/1
1st failure: https://github.com/zgoda/flask-rollup/runs/2356657471?check_suite_focus=true
main branch update: https://github.com/zgoda/flask-rollup/runs/2356670360?check_suite_focus=true
failure after PR rebase: https://github.com/zgoda/flask-rollup/runs/2356679720?check_suite_focus=true

Workflow def: https://github.com/zgoda/flask-rollup/blob/master/.github/workflows/tests.yml
Coveralls project: https://coveralls.io/github/zgoda/flask-rollup

**Comment by afinetooth** [comment]
I think the cause may be a specific combination of JSON parameters in your POST to the [Coveralls API](https://docs.coveralls.io/api-reference) that only happens in this particular scenario.

That scenario is when [coveralls-python](https://github.com/TheKevJames/coveralls-python) encounters an initial `422` from our API server and re-submits the request, coming up with a new value for `service_job_id`. (Which happens in [this line of code](https://github.com/TheKevJames/coveralls-python/blob/master/coveralls/api.py#L234).)

The particulars are discussed in [this known issue](https://github.com/TheKevJames/coveralls-python/blob/405e90f1e7daa5a61a3172fd33e5ecc6d25fc7c5/docs/usage/configuration.rst#github-actions-support) surfaced in the [repo docs here](https://github.com/TheKevJames/coveralls-python/blob/405e90f1e7daa5a61a3172fd33e5ecc6d25fc7c5/docs/usage/configuration.rst#github-actions-support).

> Sometimes Github Actions gets a little picky about the service name which needs to be used in various cases. If you run into issues, try setting the `COVERALLS_SERVICE_NAME` explicitly to either `github` or `github-actions`. It seems to be the case that you should use the `--service=github` value if you are also planning to use the `GITHUB_TOKEN` env var, and `github-actions` (which is the default) in any other case, but we've have conflicting reports on this: YMMV! See [#452](https://github.com/TheKevJames/coveralls-python/issues/252) for more info.

And in the [issue mentioned there (#452)](https://github.com/TheKevJames/coveralls-python/issues/252).

It's pernicious because the scenario is obscure, but I think the root may be the combination of `service_job_id` being set to `None` ([the second time around](https://github.com/TheKevJames/coveralls-python/blob/master/coveralls/api.py#L234)) when `service_name` is set to `github-actions`. 

Not 100% sure. Still looking through our code to see if I can tell which combination could make our back-end punt. 

I will try to develop a theory on that, but in the meantime, I'd like to see some (more) debug output in your build log. Specifically, the JSON you're sending to the [Coveralls API](https://docs.coveralls.io/api-reference) in that particular request (from your step (5) above).

Have you looked at the debug output given by `coveralls debug` per the [coveralls-python docs](https://coveralls-python.readthedocs.io/en/latest/troubleshooting.html?highlight=debug#troubleshooting)?

Could you please try that and see if it outputs the JSON being packaged up for the POST?

When this pattern happens again, I'd like to know what you're sending for:

- `service_name` (`github` or `github-actions`)
- `service_job_id`
- `repo_token`

**Comment by zgoda** [comment]
Excerpt from JSON data sent after PR rebase run as logged by Actions using defaults:

```
"service_job_id": null,
"repo_token": "[secure]",
"service_name": "github-actions",
```

If I explicitly set service name to "github" then sending coverage fails for regular push jobs and succeeds for PR rebases (exactly opposite as using defaults).

I see that succesfully submitted jobs also send `"service_job_id": null`.

**Comment by afinetooth** [comment]
Thanks, @zgoda. This is helpful.

I'll share some background below, but start with the recommendation for tl;dr:

### Fix
You might be able to avoid the failure in Step (5) above altogether by starting with:

```
# .coveralls.yml
service_name: github-actions
```

And:

```
# workflow.yml
env:
    - name: Coveralls report
      if: matrix.python-version == 3.8
      env:
        COVERALLS_REPO_TOKEN: ${{ secrets.COVERALLS_REPO_TOKEN }}
        GITHUB_TOKEN: $COVERALLS_REPO_TOKEN
      run: |
        python -m pip install -U coveralls
        coveralls -v --service=github-actions
```

_Note: Perhaps a little redundant to call out the service name twice, in `.coveralls.yml` and your workflow yaml, but it won't hurt._

---

If that doesn't avoid the issue, then we'll need to submit a change request to [coveralls-python](https://github.com/TheKevJames/coveralls-python), which [I'll start in a comment on this issue](https://github.com/TheKevJames/coveralls-python/issues/252), which is:

> Based on current code [here](https://github.com/TheKevJames/coveralls-python/blob/master/coveralls/api.py#L234), if the `service_job_id` parameter will be `None`/`null`, then the `service_name` parameter will have to be `github-actions` and the `repo_token` parameter will have to be the `coveralls repo token` provided on the repo's Coveralls Repo Page. This value is the same as what Coveralls would have you pass as `COVERALLS_REPO_TOKEN`, and means that, in this scenario, the `GITHUB_TOKEN` env var in the workflow file should be set to the `coveralls repo token`, as opposed to the `${{ secrets.GITHUB_TOKEN }}`, as currently prescribed [here](https://github.com/TheKevJames/coveralls-python/blob/405e90f1e7daa5a61a3172fd33e5ecc6d25fc7c5/docs/usage/configuration.rst#github-actions-support).

> All together:

```
# incoming JSON params

"service_name": "github-actions"
"service_job_id": <Anything, including None/NULL>
"repo_token": <Value of coveralls repo token provided by Coveralls; same as COVERALLS_REPO_TOKEN>
```

> Alternately, since there is a conditional [here](https://github.com/TheKevJames/coveralls-python/blob/master/coveralls/api.py#L234) that can result in `service_job_id` being `non-NULL`, in this case, `service_name` _could_ be set to `github`, in which case, `repo_token` would _have_ to be set to `${{ secrets.GITHUB_TOKEN }}`. 

> All together:

```
# incoming JSON params

"service_name": "github"
"service_job_id": <Some non-null value; the Coveralls Github Action sends $GITHUB_RUN_ID>
"repo_token": <Value of ${{ secrets.GITHUB_TOKEN }}>
```

> But we'd probably recommend keeping it simple and sticking with `service_name: github-actions` and `repo_token: <coveralls repo token>`, vis-a-vis env var `GITHUB_TOKEN` in workflow file. (As per the solution recommended [here](https://github.com/lemurheavy/coveralls-public/issues/1435#issuecomment-823508409).)

> Background on these suggestions below:

## Background:

The [Coveralls API](https://docs.coveralls.io/api-reference)---specifically the `/jobs` endpoint, described in our [API Reference here](https://docs.coveralls.io/api-reference), expects that, when the incoming `service_name` parameter is `github`, it is receiving a job processed by the [Coveralls Github Action](https://github.com/marketplace/actions/coveralls-github-action). 

As part of that, it expects a non-null value for `service_job_id` and a value for `repo_token` equal to the `${{ secrets.GITHUB_TOKEN }}` for the given repo.

All together:

```
# incoming JSON params

"service_name": "github"
"service_job_id": <Some non-null value; the Coveralls Github Action sends $GITHUB_RUN_ID>
"repo_token": <Value of ${{ secrets.GITHUB_TOKEN }}>
```

On the other hand, the [Coveralls API](https://docs.coveralls.io/api-reference) does _not_ have a specific way of handling an incoming job whose `service_name` is `github-actions`. It treats these jobs the same way it treats a job coming from a generic CI, which means that it expects `repo_token` to equal the `coveralls repo token` provided on the Coveralls Repo Page, and can accept `null` for `service_job_id`.

All together:

```
# incoming JSON params

"service_name": "github-actions"
"service_job_id": <Anything, including NULL>
"repo_token": <Value of coveralls repo token provided by Coveralls; same as COVERALLS_REPO_TOKEN>
```

---

Hope that helps, @zgoda.

If you see something different from expected behavior, let me know.

**Comment by zgoda** [comment]
Unfortunately this did not help at all, that PR job is still failing.

**Comment by afinetooth** [comment]
Sorry to hear that. Looking at your commits. Will reply ASAP.

**Comment by afinetooth** [comment]
@zgoda this is now tied up with my response to [this comment](https://github.com/TheKevJames/coveralls-python/issues/252#issuecomment-823558454) on [this issue](https://github.com/TheKevJames/coveralls-python/issues/252), to which you've also been party.

I'll need to clone your repo and submit some builds to my development environment to track down a root cause and propose a solution, which will probably take a couple of days.

**Comment by zgoda** [comment]
Okay, if you need me to do anything to help just let me know. I suppose the issue is generic in that any build of PR after rebase would trigger the error.

**Comment by afinetooth** [comment]
Okay, @zgoda, I can't reproduce the error with [my fork of your repo](https://github.com/afinetooth/flask-rollup):
https://github.com/afinetooth/flask-rollup

Steps:

1. I changed my [.github/workflows/test.yml](https://github.com/afinetooth/flask-rollup/blob/master/.github/workflows/tests.yml) in a branch and created a [PR](https://github.com/afinetooth/flask-rollup/pull/1), 
2. I merged [that PR](https://github.com/afinetooth/flask-rollup/pull/1) into [master](https://github.com/afinetooth/flask-rollup) 
3. I cut a [new PR for recent dependabot change](https://github.com/afinetooth/flask-rollup/pull/2): on branch [dependabot/github_actions/actions/cache-v2.1.5](https://github.com/afinetooth/flask-rollup/tree/dependabot/github_actions%2Factions%2Fcache-v2.1.5)
4. I merged [master]([master](https://github.com/afinetooth/flask-rollup)) into the [dependabot PR](https://github.com/afinetooth/flask-rollup/pull/2) and observed the coveralls server log.

Each of the above builds succeeded against my development copy of coveralls, which is identical (code-wise) to production, using [coveralls-python](https://github.com/TheKevJames/coveralls-python) as the integration via Github Actions for CI.

Here's a series of screenshots related to step (4):

**Merging master into dependabot PR:**
![Screen Shot 2021-04-27 at 4 39 58 PM](https://user-images.githubusercontent.com/918667/116326029-5d68de80-a778-11eb-9ded-71956eef15f4.png)

**dependabot PR:**
![Screen Shot 2021-04-27 at 4 40 12 PM](https://user-images.githubusercontent.com/918667/116326043-648fec80-a778-11eb-97c2-e1616233cbbc.png)

**Coveralls server log, showing JSON request as received by server:**
![Screen Shot 2021-04-27 at 4 40 40 PM](https://user-images.githubusercontent.com/918667/116326058-6bb6fa80-a778-11eb-955f-d152cd72746b.png)

**dependabot PR as last commit passes CI:**
![Screen Shot 2021-04-27 at 4 40 47 PM](https://user-images.githubusercontent.com/918667/116326164-9f922000-a778-11eb-8292-788afea1b36e.png)

**Actions history, all workflows:**
![Screen Shot 2021-04-27 at 4 41 07 PM](https://user-images.githubusercontent.com/918667/116326220-c3556600-a778-11eb-9683-ff037a6dba79.png)

This is the behavior I expected given my explanation of how coveralls treats incoming POSTs to `/api/v1/jobs` with `service_name: github-actions` vs `service_name: github`.

I did not experience any 422 errors, let alone the double 422 error [where coveralls-python re-submits the build after the initial 422](https://github.com/TheKevJames/coveralls-python/blob/54be7545f6d8e71e3d82076e5d31475d4f66b463/coveralls/api.py#L234), and I believe it's because I configured my [.github/workflows/test.yml](https://github.com/afinetooth/flask-rollup/blob/master/.github/workflows/tests.yml) in the way I suggested above, ensuring `service_name: github-actions` and `GITHUB_TOKEN = COVERALLS_REPO_TOKEN = secrets.COVERALLS_REPO_TOKEN`:

```
# .github/workflows/test.yml

[...]

    - name: Coveralls report
      if: matrix.python-version == 3.8
      env:
        COVERALLS_HOST: https://3d06fff8dca5.ngrok.io
        COVERALLS_REPO_TOKEN: ${{ secrets.COVERALLS_REPO_TOKEN }}
        GITHUB_TOKEN: $COVERALLS_REPO_TOKEN
      run: |
        python -m pip install -U PyYAML coveralls
        coveralls -v --service=github-actions
```

---

Let me know if you think I've covered it, or missed something, and if there are other steps you want me to try.

**Comment by zgoda** [comment]
I think this differs from my case:

1. branch from master and create PR <- job succeeds
2. update master <- job succeeds
3. rebase PR branch on top of master (**not** merge master into PR branch) <- job fails

Dependabot automatically rebases its PRs after discovering changes in main branch, this may be forced by issuing `@dependabot rebase` comment.

**Comment by afinetooth** [comment]
@zgoda got it. I misunderstood the use case, apologies. I'll try that and get back asap.

**Comment by afinetooth** [comment]
Ok, @zgoda, I've done this three (3) times without being able to reproduce the 422 error.

Here's the first rebased PR:
https://github.com/afinetooth/flask-rollup/pull/4

And its corresponding Action run:
https://github.com/afinetooth/flask-rollup/runs/2479104801

The second rebased PR:
https://github.com/afinetooth/flask-rollup/pull/5

Its Action run:
https://github.com/afinetooth/flask-rollup/runs/2479265562

And the third rebased PR (still open):
https://github.com/afinetooth/flask-rollup/pull/6

And its Action run:
https://github.com/afinetooth/flask-rollup/runs/2479348854?check_suite_focus=true

Note that I did this a 2nd and 3rd time because it's been a while since I rebased a local branch after already pushing some commits to remote (to open the PR). The 1st and 2nd times I felt unsure about the rebase action when I realized they used the "[merge strategy](https://www.atlassian.com/git/tutorials/using-branches/merge-strategy#:~:text=Recursive%20is%20the%20default%20merge,pulling%20or%20merging%20one%20branch)" of "recursive merge" (the _default_ strategy [according to the man page](http://man.he.net/man1/git-rebase)). Had not seen that before, so I think I was used to local rebases with no complaints (I assume, "fast-forward merge strategy"?). For the 3rd rebase, I followed [the advice here](http://actimem.com/git/merge/) (under "Pushing feature branch after Rebase") and simply force-pushed. I'm not sure if this changed the merge strategy, but there were no complaints from git and this seemed the cleanest. Again, you can review the [open PR here](https://github.com/afinetooth/flask-rollup/pull/6) and compare it to the others [ [1st PR](https://github.com/afinetooth/flask-rollup/pull/4) | [2nd PR](https://github.com/afinetooth/flask-rollup/pull/5) ] if it's helpful.

**Comment by duttonw** [comment]
Github actions has had major changes to security due to this https://securitylab.github.com/research/github-actions-preventing-pwn-requests/ and https://securitylab.github.com/research/github-actions-untrusted-input/

dependabot got hit massively 
https://github.com/dependabot/dependabot-core/issues/3253

**Comment by afinetooth** [comment]
Thanks, @duttonw. For those interested, this seems to be the canonical approach to modifying your workflow for secure use of dependabot, including auto-merge:
https://github.com/dependabot/dependabot-core/issues/3253#issuecomment-797125425

**Comment by pbelskiy** [comment]
So, any update here? I've got this error on PR, but on master not.

https://github.com/pbelskiy/aiojenkins/pull/23

**Comment by afinetooth** [comment]
@pbelskiy can I pls ask you to create a new issue with your specific issue, coveralls repo URL, and build URL(s) that are affected? I assume you're getting a 422 error, but I'm thrown off by the history here and the number of users / different possible scenarios. Feel free to refer to [this issue](https://github.com/lemurheavy/coveralls-public/issues/1435) or any relevant comment here. Thanks.

**Comment by hartwork** [comment]
Ran [into this](https://github.com/asyncee/django-easy-select2/runs/4619678877?check_suite_focus=true) myself today (in context of https://github.com/asyncee/django-easy-select2/pull/96); `--service=github` proved to work around the issue. To me that means that likely auto-detection of GitHub Actions has issues.

I have also seen coveralls crash when there is no `GITHUB_REF` in the environment, which happens easily when coveralls is invoked by tox. I have created a dedicated ticket #1607 for that now.

**Comment by stephane-caron** [comment]
Had this problem again today. The following workaround fixed it for me, sharing it for future travelers:

* Sign into Coveralls via GitHub
* Go to [Add Repos](https://coveralls.io/repos/new)
* Un-tick then re-tick the failing project
* Re-run the failing CI jobs

Here is the corresponding workflow: [`main.yml`](https://github.com/stephane-caron/loop-rate-limiters/blob/6074a92ede3b85abd44542f19a8a763710780791/.github/workflows/main.yml).

**Comment by Krinkle** [comment]
💩
